File: Design Pattern/Creational/Prototype/shape.ts
interface Cloneable {
    clone(): Cloneable;
}

class Circle implements Cloneable {
    color: string;
    radius: number;
    constructor(color: string, radius: number) {
        this.color = color;
        this.radius = radius;
    }

    // TODO: Implement clone()
    clone() : Cloneable {
        return new Circle(this.color, this.radius);
    }

    printInfo(): void {
        console.log(`Circle [Color: ${this.color}, Radius: ${this.radius.toFixed(1)}]`);
    }
}

class Rectangle implements Cloneable {
    color: string;
    width: number;
    height: number;
    constructor(color: string, width: number, height: number) {
        this.color = color;
        this.width = width;
        this.height = height;
    }

    // TODO: Implement clone()
    clone() : Cloneable {
        return new Rectangle(this.color, this.width, this.height);
    }

    printInfo(): void {
        console.log(`Rectangle [Color: ${this.color}, Width: ${this.width.toFixed(1)}, ` +
                     `Height: ${this.height.toFixed(1)}]`);
    }
}

// After implementing:
const original = new Circle("Red", 5.0);
const cloned = original.clone() as Circle;
cloned.radius = 10.0;
original.printInfo();
cloned.printInfo();

const rect = new Rectangle("Blue", 4.0, 6.0);
const clonedRect = rect.clone() as Rectangle;
clonedRect.width = 8.0;
rect.printInfo();
clonedRect.printInfo();

File: Design Pattern/Creational/Prototype/shape.cpp
#include <iomanip>
#include <iostream>
#include <string>
using namespace std;

class Cloneable {
public:
  virtual Cloneable *clone() const = 0;
  virtual ~Cloneable() = default;
};

class Circle : public Cloneable {
private:
  string color;
  double radius;

public:
  Circle(string color, double radius) : color(move(color)), radius(radius) {}

  // TODO: Implement clone()
  Circle *clone() const override { return new Circle(color, radius); }

  void setRadius(double r) { radius = r; }

  void printInfo() const {
    cout << std::fixed << std::setprecision(1);
    cout << "Circle [Color: " << color << ", Radius: " << radius << "]" << endl;
  }
};

class Rectangle : public Cloneable {
private:
  string color;
  double width, height;

public:
  Rectangle(string color, double w, double h)
      : color(move(color)), width(w), height(h) {}

  // TODO: Implement clone()
  Rectangle *clone() const override {
    return new Rectangle(color, width, height);
  }

  void setWidth(double w) { width = w; }

  void printInfo() const {
    cout << std::fixed << std::setprecision(1);
    cout << "Rectangle [Color: " << color << ", Width: " << width
         << ", Height: " << height << "]" << endl;
  }
};

int main() {
  // After implementing, usage should look like:
  Circle original("Red", 5.0);
  Circle *cloned = static_cast<Circle *>(original.clone());
  cloned->setRadius(10.0);
  original.printInfo(); // Circle [Color: Red, Radius: 5.0]
  cloned->printInfo();  // Circle [Color: Red, Radius: 10.0]
  delete cloned;

  Rectangle rect("Blue", 4.0, 6.0);
  Rectangle *clonedRect = static_cast<Rectangle *>(rect.clone());
  clonedRect->setWidth(8.0);
  rect.printInfo();        // Rectangle [Color: Blue, Width: 4.0, Height: 6.0]
  clonedRect->printInfo(); // Rectangle [Color: Blue, Width: 8.0, Height: 6.0]
  delete clonedRect;
  return 0;
}
